### PR TITLE
Convert pytest.py into a package

### PR DESCRIPTION
**Pull request by nicoddemus**
As discussed in https://github.com/pytest-dev/pytest/issues/3342, this is the first step to make pytest fully support static typing.

Had a few minutes on my hands so decide to tackle this.

### CODE REVIEW

**Comment by bluetech** [comment]
My PR from yesterday https://github.com/pytest-dev/pytest/pull/6302 is pretty much the same. It failed CI twice on the Azure py37 job, timing out on the `testing/test_assertion.py::test_exception_handling_no_traceback` test. It *seems* like this one is going down the same path. I haven't looked but not sure how this affects this test, and it doesn't reproduce locally.

**Comment by nicoddemus** [comment]
Oh really sorry @bluetech, I've completely missed the notification on that (in fact I still can't see it in my Inbox, weird), otherwise I wouldn't have opened this one. 😓

**Comment by CaselIT** [comment]
Sorry if already discusses somewhere, the long term plan is to remove the `_pytest` package in favor of this one?

**Comment by nicoddemus** [comment]
> Sorry if already discusses somewhere, the long term plan is to remove the _pytest package in favor of this one?

No, the `pytest` package contains only symbols of the public API, while `_pytest` contains the actual codebase.

**Comment by nicoddemus** [comment]
It is incredible how it is basically impossible to restart an Azure job... Clicking on `Rebuild` on the menu below does absolutely nothing:

image

File: src/pytest/__main__.py [hunk 1]
```diff
@@ -0,0 +1,6 @@
+"""
+pytest entry point
+"""
+import pytest
+
```
Comment (RonnyPfannschmidt) [inline src/pytest/__main__.py]: i vaguely recall its necessary to check for `__name__ == "__main__"` to guard against certain user mistakes

**Comment by asottile** [comment]
the failure is quite weird, I can't reproduce it w/ 3.7.5x64 locally running `tox -e py37-twisted-numpy` :thinking:

File: src/pytest/__main__.py (hunk 1 above)
Comment (nicoddemus) [reply src/pytest/__main__.py]: OK, will update it later, thanks

**Comment by nicoddemus** [comment]
> the failure is quite weird, I can't reproduce it w/ 3.7.5x64 locally running

Me neither... just pushed a skip for that test to see if it is that test the culprit, or something else.

**Comment by nicoddemus** [comment]
Weird, it is just that test who seems to be stucking CI... 🤔

**Comment by nicoddemus** [comment]
Rá looking at the test things are clearer: the test uses `multiprocessing`, so having the `__main__` guard as @RonnyPfannschmidt pointed out is mandatory. 😁

**Pull request merged**